app/models/client/models/ChatMessage.js

```diff
@@ -9,3 +27,101 @@ ChatMessage.setReactions = function(messageId, reactions) {
 ChatMessage.unsetReactions = function(messageId) {
 	return this.update({ _id: messageId }, { $unset: { reactions: 1 } });
 };
+
+const normalizeThreadMessage = (message) => {
+	if (message.msg) {
+		return renderMessageBody(message).replace(/<br\s?\\?>/g, ' ');
+	}
+
+	if (message.attachments) {
+		const attachment = message.attachments.find((attachment) => attachment.title || attachment.description);
+
+		if (attachment && attachment.description) {
+			return s.escapeHTML(attachment.description);
+		}
+
+		if (attachment && attachment.title) {
+			return s.escapeHTML(attachment.title);
+		}
+	}
+};
+
+const upsertMessage = async ({ msg, subscription, uid = Tracker.nonreactive(() => Meteor.userId()) }, collection = ChatMessage) => {
+	const userId = msg.u && msg.u._id;
+
+	if (subscription && subscription.ignored && subscription.ignored.indexOf(userId) > -1) {
+		msg.ignored = true;
+	}
+
+	if (msg.t === 'e2e' && !msg.file) {
+		msg.e2e = 'pending';
+	}
+	msg = await promises.run('onClientMessageReceived', msg) || msg;
+
+	const { _id, ...messageToUpsert } = msg;
+
+	if (msg.tcount) {
+		collection.direct.update({ tmid: _id }, {
+			$set: {
+				following: msg.replies && msg.replies.indexOf(uid) > -1,
+				threadMsg: normalizeThreadMessage(messageToUpsert),
+				repliesCount: msg.tcount,
+			},
+		}, { multi: true });
+	}
+
+	return collection.direct.upsert({ _id }, messageToUpsert);
+};
+
+function upsertMessageBulk({ msgs, subscription }, collection = ChatMessage) {
+	const uid = Tracker.nonreactive(() => Meteor.userId());
+	const { queries } = ChatMessage;
+	collection.queries = [];
+	msgs.forEach((msg, index) => {
+		if (index === msgs.length - 1) {
+			ChatMessage.queries = queries;
+		}
+		upsertMessage({ msg, subscription, uid }, collection);
+	});
+}
+
+const messagePreFetch = () => {
+	let messagesFetched = false;
+	Tracker.autorun(() => {
+		if (!messagesFetched && CachedChatSubscription.ready.get()) {
+			const status = Meteor.status();
+			if (status.status !== 'connected') {
+				return;
+			}
+			messagesFetched = true;
+			const subscriptions = ChatSubscription.find(
+				{
+					open: true,
+				},
+				{
+					fields: {
+						rid: 1,
+						ls: 1,
+					},
+				},
+			);
+			const limit = parseInt(getConfig('roomListLimit')) || 50;
+			subscriptions.forEach((subscription) => {
+				const ts = undefined;
+				const { rid, ls } = subscription;
+				Meteor.call('loadHistory', rid, ts, limit, ls, (err, result) => {
+					if (err) {
+						return;
+					}
+					const { messages = [] } = result;
+					upsertMessageBulk({
+						msgs: messages.filter((msg) => msg.t !== 'command'),
+						subscription,
+					});
+				});
+			});
+		}
+	});
+};
+
+callbacks.add('afterMainReady', messagePreFetch, callbacks.priority.LOW, 'messagePreFetch');
```

Comment:
The prefetch needs to be optional and disabled by default as a beta feature, I don't know how bad it can be for users with a huge number of rooms like me (> 1000)

Reply:
This prefetch was done keeping PWA experience in mind. If the user opens the app without internet connection then the rooms which were not accessed by the user won't load any message

Reply:
I see, but it should be a setting keeping the performance in mind, we can't force every installation to start prefetching hundred or thousand of rooms if we don't know if it would cause performance issues, mainly for big installations or heavy users.

Reply:
Ok sure, will make it configurable